blogposts/announcing-sourcegraph-3.11.md

```diff
@@ -0,0 +1,196 @@
+
+Sourcegraph Automation campaigns are in private beta. [Watch the Automation screencasts](https://about.sourcegraph.com/product/automation#see-it-in-action) to see what we have planned and [apply for early access](https://about.sourcegraph.com/contact/request-automation-demo/) to Automation for your organization.
+
+## Search query language statistics
+
+<p class="container">
+  <div style="padding:56.25% 0 0 0;position:relative;">
+    <iframe src="https://player.vimeo.com/video/380662321?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
+  </div>
+  <p style="text-align: center"><a href="https://vimeo.com/380662321" target="_blank">View on Vimeo</a></p>
+</p>
+
+A new experimental feature has been added to show [language statistics about your search query](https://docs.sourcegraph.com/user/search#statistics). Search result pages now have a **Stats** link to a visual breakdown of the languages that comprise the results for the query. This data is also available through our [GraphQL API](https://docs.sourcegraph.com/api/graphql).
+
+Language analysis is computationally expensive, so this feature is currently behind a feature flag. To enable, update your global, organization, or user settings to include `{ “experimentalFeatures”: { “searchStats”: true } }`.
+
+## Introducing code aware structural search
+
+<p class="container">
+  <div style="padding:56.25% 0 0 0;position:relative;">
+    <iframe src="https://player.vimeo.com/video/380662673?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
+  </div>
+  <p style="text-align: center"><a href="https://vimeo.com/380662673" target="_blank">View on Vimeo</a></p>
+</p>
+
+Sourcegraph 3.11 introduces structural search, a code aware search syntax that can identify structural patterns in code. This enables advanced code search to specifically match patterns inside code structures such as function parameters and loop bodies.
+
+It can be awkward or difficult to match code blocks or nested expressions with regexp. To meet this challenge we’ve introduced a new and easier way to search code that operates more closely on the parse tree of the input using [Comby syntax](https://comby.dev/) for structural matching.
+
+Here is a search query that [matches all the arguments in `fmt.Sprintf` calls](https://sourcegraph.com/search?q=repo%3A%5Egithub%5C.com%2Fsourcegraph%2Fsourcegraph%24++%27fmt.Sprintf%28%3A%5Bargs%5D%29%27+count%3A100&patternType=structural) in the Sourcegraph code base:
+
+```text
+patterntype:structural “fmt.Sprintf(:[args])”
```

Comment:
I don't think this is a good example. For interactive searching, this is not really better than just searching for `fmt.Sprintf(`. We should use an example that shows off the benefits of structural search. For example:

- https://sourcegraph.com/search?q=repo:sourcegraph/sourcegraph%24+fmt.Sprintf%28%22:%5Bstr%5D%22%29&patternType=structural - matches unnecessary usage of `fmt.Sprintf`
- https://sourcegraph.com/search?q=repo:sourcegraph/sourcegraph%24+%22http.Client%7B:%5Ba%5D+Transport:+:%5Bb%5D+:%5Bc%5D%7D%22+count:1000&patternType=structural all places where we make a http.Client with a Transport (spans multiple lines and can be in any order, so structural search shines here)
- https://sourcegraph.com/search?q=repo:sourcegraph/sourcegraph%24+%22make%28:%5Bx%5D%2C+0%29%22+count:1000&patternType=structural places where we could have just done `x := []type{}` (fewer chars, equivalent, see https://stackoverflow.com/questions/29164375/correct-way-to-initialize-empty-slice)
- https://sourcegraph.com/search?q=repo:sourcegraph/sourcegraph%24+%22func%28:%5B%5Bvarname%5D%5D+:%5Btypes%5D%29+%7B%7D%22+count:1000&patternType=structural places where we could remove the var names in the param list (eg `func(w http.ResponseWriter) {}` -> `func(http.ResponseWriter) {}`)

Also, let's resolve https://github.com/sourcegraph/sourcegraph/issues/7330 before we include this, or mention limitations/known issues (that tripped me up).

cc @rvantonder

Reply:
#7330: this is expected behavior (the repo searched is unindexed). That limitation is mentioned in the [user docs](https://docs.sourcegraph.com/user/search/structural#current-functionality-and-restrictions), but making this obvious could be improved. Will track updates/solutions in that issue.

Re example: in the docs I start off with a super simple example just to introduce syntax. We don't have as much space in the blog post, so it can cut to the chase as far as the example goes.

Reply:
I updated with new examples and added limitations. Let me know what you think!

Reply:
I'll look into adding an alert message unindexed search soon on sourcegraph.com so that it isn't a confusing/surprising behavior.